crates/menmos-client/src/metadata_detector.rs

```diff
@@ -0,0 +1,68 @@
+use std::{collections::HashMap, path::Path};
+
+use serde_json;
+use snafu::{ResultExt, Snafu};
+
+use crate::Meta;
+
+#[derive(Debug, Snafu)]
+pub enum MetadataDetectorError {
+    #[snafu(display("failed to load the mime types: {}", source))]
+    MimeTypesDeserializationError { source: serde_json::Error },
+}
+
+type Result<T> = std::result::Result<T, MetadataDetectorError>;
+
+#[derive(Clone)]
+pub struct MetadataDetector {
+    mime_types: HashMap<String, String>,
+}
+
+impl MetadataDetector {
+    pub fn new() -> Result<Self> {
+        let bytes = include_bytes!("data/mime-types.json");
+
+        let mime_types = serde_json::from_slice(bytes).context(MimeTypesDeserializationError)?;
+
+        Ok(Self { mime_types })
+    }
+
+    fn detect_mime_type<P: AsRef<Path>>(&self, path: P) -> Option<&String> {
+        let ext = path.as_ref().extension()?;
+        let ext_str = ext.to_string_lossy().to_string();
+
+        let mime_type = self.mime_types.get(&ext_str)?;
+
+        Some(mime_type)
+    }
+
+    pub fn populate<P: AsRef<Path>>(&self, path: P, meta: &mut Meta) -> Result<()> {
+        if let Some(mime_type) = self.detect_mime_type(path) {
+            meta.metadata
+                .insert(String::from("extension"), mime_type.clone());
+        }
+
+        Ok(())
+    }
+}
+
+#[test]
+fn detect_file_mime_type() {
+    let path = Path::new("foo.txt");
```

Comment:
Since your `detect_mime_type` method takes a `P: AsRef<Path>` you don't really need to create a Path here, because `&str` implements `AsRef<Path>`. You can simply use your `&str` directly! 😄